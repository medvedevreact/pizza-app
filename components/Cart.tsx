// Cart.tsx
"use client";
import React from "react";
import {
  Sheet,
  SheetContent,
  SheetHeader,
  SheetTitle,
  SheetTrigger,
  SheetFooter,
} from "@/components/ui/sheet";
import { useCartStore } from "@/store/cart";
import { Button } from "./ui/button";
import { useUserStore } from "@/store/user";
import { useRouter } from "next/navigation";
import { CartItem } from "./СartItem";
import toast from "react-hot-toast";

interface CartProps {
  children: React.ReactNode;
}

export const Cart: React.FC<CartProps> = ({ children }) => {
  const { cartItems, totalPrice } = useCartStore((state) => state);
  const user = useUserStore((state) => state.user);
  const router = useRouter();

  const moveToOrderPage = () => {
    if (user && user.emailVerified) {
      router.push("/checkout");
    } else if (user && !user.emailVerified) {
      toast.error(
        "Подтвердите электронную почту. Если вы уже подтвердили, пожалуйста, обновите страницу, чтобы перейти к оформлению заказа."
      );
    } else {
      toast.error(
        "Вы не авторизованы. Авторизуйтесь, чтобы перейти к оформлению заказа."
      );
    }
  };

  return (
    <Sheet>
      <SheetTrigger asChild>{children}</SheetTrigger>
      <SheetContent className="p-4 flex flex-col h-full">
        <SheetHeader>
          <SheetTitle
            hidden={cartItems.length === 0}
            className="mb-4 text-lg sm:text-xl"
          >
            Ваша корзина:
          </SheetTitle>
        </SheetHeader>
        <div className="flex-grow overflow-y-auto space-y-4">
          {cartItems.length === 0 ? (
            <div className="flex flex-col items-center justify-center h-full">
              <img src="/img/emptyCartIcon.png" className="mb-4" alt="" />
              <p className="text-lg font-semibold text-center">
                Ваша корзина пуста. Пожалуйста, добавьте товары, чтобы
                продолжить.
              </p>
            </div>
          ) : (
            cartItems.map((item) => <CartItem key={item.id} item={item} />)
          )}
        </div>
        {cartItems.length != 0 && (
          <SheetFooter className="mt-6">
            <div className="flex flex-col w-full">
              <div className="flex justify-between items-center w-full mb-4">
                <p className="text-lg font-semibold">Сумма заказа:</p>
                <p className="text-lg font-semibold">{totalPrice} ₽</p>
              </div>
              <Button variant="destructive" onClick={() => moveToOrderPage()}>
                Оформить заказ
              </Button>
            </div>
          </SheetFooter>
        )}
      </SheetContent>
    </Sheet>
  );
};
