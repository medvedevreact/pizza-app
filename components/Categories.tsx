"use client";
import React from "react";
import { cn } from "@/lib/utils";
import { Container } from "./Container";
import { useCategoryStore } from "@/store/category";

interface CategoriesProps {
  className?: string;
  categories: {
    id: number;
    title: string;
    created_at: Date;
  }[];
}

export const Categories: React.FC<CategoriesProps> = ({
  className,
  categories,
}) => {
  const activeCategory = useCategoryStore((state) => state.activeCategory);

  return (
    <div className={cn("h-[75px] flex items-center bg-white", className)}>
      <Container className="flex justify-start w-full">
        <ul className="flex flex-wrap gap-[60px]">
          {categories.map((category) => (
            <li key={category.id}>
              <a
                href={`#${category.title}`}
                className={`cursor-pointer  py-2 ${
                  category.title === activeCategory ? "text-orange-500" : ""
                }`}
              >
                {category.title}
              </a>
            </li>
          ))}
        </ul>
      </Container>
    </div>
  );
};
