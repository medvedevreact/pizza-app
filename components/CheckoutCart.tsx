"use client";
import React from "react";

import { WhiteBlock } from "./WhiteBlock";
import { useCartStore } from "@/store/cart";
import { CartItem } from "./СartItem";

interface CheckoutCartProps {
  className?: string;
}

export const CheckoutCart: React.FC<CheckoutCartProps> = ({}) => {
  const cartItems = useCartStore((state) => state.cartItems);

  return (
    <WhiteBlock className="mb-5" title="1. Корзина">
      <ul className="flex flex-col gap-3 max-h-[500px] overflow-y-auto p-2">
        {cartItems.map((item) => (
          <CartItem key={item.id} item={item} />
        ))}
      </ul>
    </WhiteBlock>
  );
};
