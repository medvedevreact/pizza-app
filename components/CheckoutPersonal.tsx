import React from "react";
import { useFormContext } from "react-hook-form";
import { WhiteBlock } from "./WhiteBlock";
import { Label } from "./ui/label";
import { Input } from "./ui/input";
import { Textarea } from "./ui/textarea";

export const CheckoutPersonal = () => {
  const {
    register,
    formState: { errors },
  } = useFormContext();

  return (
    <WhiteBlock title="2. Персональные данные">
      <div className="grid md:grid-cols-2 gap-4 grid-cols-1">
        <div>
          <Label htmlFor="firstname">Имя</Label>
          <Input
            className="mb-2"
            id="firstname"
            placeholder="Введите ваше имя..."
            {...register("firstname")}
          />
          {errors.firstname && (
            <span className="text-red-600">
              {errors.firstname.message as string}
            </span>
          )}
        </div>
        <div>
          <Label htmlFor="lastname">Фамилия</Label>
          <Input
            className="mb-2"
            id="lastname"
            placeholder="Введите вашу фамилию..."
            {...register("lastname")}
          />
          {errors.lastname && (
            <span className="text-red-600">
              {errors.lastname.message as string}
            </span>
          )}
        </div>
        <div>
          <Label htmlFor="phone">Телефон</Label>
          <Input
            className="mb-2"
            id="phone"
            placeholder="Введите ваш телефон..."
            {...register("phone")}
          />
          {errors.phone && (
            <span className="text-red-600 ">
              {errors.phone.message as string}
            </span>
          )}
        </div>
        <div>
          <Label htmlFor="phone">Комментарий к заказу (опционально)</Label>
          <Textarea
            placeholder="Здесь вы можете добавить комментарий к заказу..."
            className="mb-2 min-h-[200px]"
            {...register("comment")}
          />
          {errors.comment && (
            <span className="text-red-600 ">
              {errors.comment.message as string}
            </span>
          )}
        </div>
      </div>
    </WhiteBlock>
  );
};
