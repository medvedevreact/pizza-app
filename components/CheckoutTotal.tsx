"use client";
import React from "react";
import { WhiteBlock } from "./WhiteBlock";
import { Button } from "@/components/ui/button";
import { Label } from "@/components/ui/label";

interface CheckoutTotalProps {
  className?: string;
  totalPrice: number;
  paymentMethod: string;
  handlePaymentChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

export const CheckoutTotal: React.FC<CheckoutTotalProps> = ({
  totalPrice,
  paymentMethod,
  handlePaymentChange,
}) => {
  return (
    <WhiteBlock className="lg:w-[380px] h-full w-full" title="Итого:">
      <div className="flex flex-col justify-between h-full">
        <div className="text-2xl font-bold mb-8">
          Итоговая цена: {totalPrice} ₽
        </div>
        <div className="mb-8">
          <Label>Выберите способ оплаты:</Label>
          <div className="mt-2 flex flex-col gap-2">
            <label
              className={`px-4 py-2 border-2 rounded-full cursor-pointer transition-colors ${
                paymentMethod === "card"
                  ? "border-orange-500 bg-orange-300"
                  : "border-gray-300"
              }`}
            >
              <input
                type="radio"
                name="paymentMethod"
                value="card"
                checked={paymentMethod === "card"}
                onChange={handlePaymentChange}
                className="hidden"
              />
              Картой на кассе
            </label>
            <label
              className={`px-4 py-2 border-2 rounded-full cursor-pointer transition-colors ${
                paymentMethod === "cash"
                  ? "border-orange-500 bg-orange-300"
                  : "border-gray-300"
              }`}
            >
              <input
                type="radio"
                name="paymentMethod"
                value="cash"
                checked={paymentMethod === "cash"}
                onChange={handlePaymentChange}
                className="hidden"
              />
              Наличными на кассе
            </label>
          </div>
        </div>
        <Button
          disabled={totalPrice === 0}
          type="submit"
          className="w-full py-2 bg-orange-500 text-white  mb-3"
        >
          Создать заказ
        </Button>
      </div>
    </WhiteBlock>
  );
};
