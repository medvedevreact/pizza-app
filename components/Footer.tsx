import React from "react";
import { cn } from "@/lib/utils";

interface FooterProps {
  className?: string;
}

export const Footer: React.FC<FooterProps> = ({ className }) => {
  return (
    <footer
      className={cn(
        "bg-[rgb(228,228,228)] text-gray-800 py-6 text-center",
        className
      )}
    >
      <div className="social-icons flex justify-center mb-4">
        <a href="#" className="mx-4 hover:text-orange-400">
          Facebook
        </a>
        <a href="#" className="mx-4 hover:text-orange-400">
          Instagram
        </a>
        <a href="#" className="mx-4 hover:text-orange-400">
          Twitter
        </a>
      </div>
      <p className="mb-2">&copy; 2025 El Pizza. Все права защищены.</p>
      <p>
        Контакты:{" "}
        <a href="mailto:info@elpizza.com" className="hover:text-orange-400">
          info@elpizza.com
        </a>{" "}
        | Телефон: +123 456 7890
      </p>
    </footer>
  );
};
