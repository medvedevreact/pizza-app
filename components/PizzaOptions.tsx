"use client";
import React from "react";
import { cn } from "@/lib/utils";

interface PizzaOptionsProps {
  className?: string;
  product: any;

  size: string;
  setSize: (size: string) => void;
  pizzaType: string;
  setPizzaType: (type: string) => void;
  selectedIngredients: any[];
  handleIngredientChange: (ingredient: any) => void;
}

export const PizzaOptions: React.FC<PizzaOptionsProps> = ({
  className,
  product,

  size,
  setSize,
  pizzaType,
  setPizzaType,
  selectedIngredients,
  handleIngredientChange,
}) => {
  return (
    <div className={cn("space-y-4 mb-10", className)}>
      <h2 className="text-lg font-semibold">Выберите размер пиццы:</h2>
      <div className="flex space-x-4">
        <label
          className={`px-4 py-2 border-2 rounded-full cursor-pointer transition-colors ${
            size === "30"
              ? "border-orange-500 bg-orange-300"
              : "border-gray-300"
          }`}
        >
          <input
            type="radio"
            name="size"
            value="30"
            checked={size === "30"}
            onChange={() => {
              setSize("30");
              setPizzaType("традиционное");
            }}
            className="hidden"
          />
          30 см
        </label>
        <label
          className={`px-4 py-2 border-2 rounded-full cursor-pointer transition-colors ${
            size === "35"
              ? "border-orange-500 bg-orange-300"
              : "border-gray-300"
          }`}
        >
          <input
            type="radio"
            name="size"
            value="35"
            checked={size === "35"}
            onChange={() => setSize("35")}
            className="hidden"
          />
          35 см
        </label>
        <label
          className={`px-4 py-2 border-2 rounded-full cursor-pointer transition-colors ${
            size === "40"
              ? "border-orange-500 bg-orange-300"
              : "border-gray-300"
          }`}
        >
          <input
            type="radio"
            name="size"
            value="40"
            checked={size === "40"}
            onChange={() => setSize("40")}
            className="hidden"
          />
          40 см
        </label>
      </div>

      <h2 className="text-lg font-semibold">Выберите тип теста:</h2>
      <div className="flex space-x-4">
        <label
          className={`px-4 py-2 border-2 rounded-full cursor-pointer transition-colors ${
            pizzaType === "традиционное"
              ? "border-orange-500 bg-orange-300"
              : "border-gray-300"
          }`}
        >
          <input
            type="radio"
            name="pizzaType"
            value="традиционное"
            checked={pizzaType === "традиционное"}
            onChange={() => setPizzaType("традиционное")}
            className="hidden"
          />
          Традиционное
        </label>
        <label
          className={`px-4 py-2 border-2 rounded-full cursor-pointer transition-colors ${
            pizzaType === "тонкое"
              ? "border-orange-500 bg-orange-300"
              : "border-gray-300"
          }`}
        >
          <input
            type="radio"
            name="pizzaType"
            value="тонкое"
            checked={pizzaType === "тонкое"}
            onChange={() => setPizzaType("тонкое")}
            className="hidden"
            disabled={size === "30"}
          />
          Тонкое
        </label>
      </div>

      <h2 className="text-lg font-semibold">Дополнительные ингредиенты:</h2>
      <div className="max-h-40 overflow-y-auto">
        <table className="min-w-full bg-white">
          <thead>
            <tr>
              <th className="py-2">Фото</th>
              <th className="py-2">Название</th>
              <th className="py-2">Цена</th>
              <th className="py-2">Добавить</th>
            </tr>
          </thead>
          <tbody>
            {product.ingredients.map((ingredient: any) => (
              <tr key={ingredient.id}>
                <td className="py-2">
                  <img
                    src={ingredient.img}
                    alt={ingredient.name}
                    className="w-10 h-10"
                  />
                </td>
                <td className="py-2">{ingredient.name}</td>
                <td className="py-2">{ingredient.price}</td>
                <td className="py-2">
                  <input
                    type="checkbox"
                    checked={selectedIngredients.some(
                      (ing) => ing.id === ingredient.id
                    )}
                    onChange={() => handleIngredientChange(ingredient)}
                    data-testid={`checkbox-${ingredient.id}`}
                  />
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    </div>
  );
};
