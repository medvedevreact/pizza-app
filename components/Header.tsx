"use client";
import React, { useEffect, useState } from "react";
import { cn } from "@/lib/utils";
import { Container } from "./Container";
import Image from "next/image";
import Logo from "../public/img/Logo.png";
import { Button } from "./ui/button";
import { SearchInput } from "./SearchInput";
import { AuthModal } from "./modals/AuthModal";
import { useUserStore } from "@/store/user";
import { onAuthStateChanged, signOut } from "firebase/auth";
import { auth } from "@/configs/firebase.config";
import { Settings, ShoppingCart, User } from "lucide-react";
import { Cart } from "./Cart";
import Link from "next/link";
import { adminList } from "@/constants/adminList";

interface HeaderProps {
  className?: string;
  hasSearch?: boolean;
  hasCart?: boolean;
  hasProfile?: boolean;
}

const defaultProps = {
  hasSearch: true,
  hasCart: true,
  hasProfile: true,
};

export const Header: React.FC<HeaderProps> = ({
  className,
  hasSearch = defaultProps.hasSearch,
  hasCart = defaultProps.hasCart,
  hasProfile = defaultProps.hasProfile,
}) => {
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [isLoading, setIsLoading] = useState(true);

  const openModal = () => setIsModalOpen(true);
  const closeModal = () => setIsModalOpen(false);

  const user = useUserStore((state) => state.user);
  const setUser = useUserStore((state) => state.setUser);

  const isAdmin = adminList.includes(String(user?.email));

  console.log(user);

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (user) => {
      setUser(user);
      setIsLoading(false);
    });

    return () => unsubscribe();
  }, [setUser]);

  const handleSignOut = () => {
    setIsLoading(true);
    signOut(auth)
      .then(() => {
        setUser(null);
        setIsLoading(false);
      })
      .catch((error) => {
        console.log("error", error);
        setIsLoading(false);
      });
  };

  return (
    <header className={cn("h-[150px]", className)}>
      <Container
        className={`flex items-center justify-between py-8 ${
          hasSearch === false ? "border-b border-gray-300" : ""
        }`}
      >
        <Link href={"/"}>
          <div className="flex items-center mr-4 cursor-pointer">
            <Image src={Logo} alt="Логотип" width={70} height={70} />
            <h1 className="md:text-3xl text-xl">El PIZZA</h1>
          </div>
        </Link>
        {hasSearch && <SearchInput />}

        {isLoading ? (
          <div className="flex items-center gap-1 cursor-pointer">
            <Button className="w-[100px]" variant="destructive" disabled>
              Загрузка...
            </Button>
            {hasCart && (
              <Cart>
                <ShoppingCart
                  data-testid="cartIcon"
                  className="text-orange-500"
                />
              </Cart>
            )}
          </div>
        ) : user ? (
          <div className="flex items-center gap-3 cursor-pointer">
            {isAdmin && (
              <Link href="/dashboard">
                <Settings className="text-orange-500 cursor-pointer" />
              </Link>
            )}
            {hasProfile && (
              <Link href="/profile">
                <User className="text-orange-500 cursor-pointer" />
              </Link>
            )}

            <Button
              className="w-[100px]"
              variant="destructive"
              onClick={handleSignOut}
            >
              Выйти
            </Button>
            {hasCart && (
              <Cart>
                <ShoppingCart
                  data-testid="cartIcon"
                  className="text-orange-500"
                />
              </Cart>
            )}
          </div>
        ) : (
          <div className="flex items-center gap-3 cursor-pointer">
            <Button
              className="w-[100px]"
              variant="destructive"
              onClick={openModal}
            >
              Войти
            </Button>
            {hasCart && (
              <Cart>
                <ShoppingCart
                  data-testid="cartIcon"
                  className="text-orange-500"
                />
              </Cart>
            )}
          </div>
        )}
      </Container>
      {isModalOpen && <AuthModal isOpen={isModalOpen} onClose={closeModal} />}
    </header>
  );
};
