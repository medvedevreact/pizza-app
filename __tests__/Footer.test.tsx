import React from "react";
import { render } from "@testing-library/react";
import { Footer } from "@/components/Footer";

describe("Footer", () => {
  it("Renders component", () => {
    const { container } = render(<Footer />);
    const footerElement = container.querySelector("footer");
    expect(footerElement).toMatchSnapshot();
  });
});
