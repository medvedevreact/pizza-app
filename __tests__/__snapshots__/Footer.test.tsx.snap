// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer Renders component 1`] = `
<footer
  class="bg-[rgb(228,228,228)] text-gray-800 py-6 text-center"
>
  <div
    class="social-icons flex justify-center mb-4"
  >
    <a
      class="mx-4 hover:text-orange-400"
      href="#"
    >
      Facebook
    </a>
    <a
      class="mx-4 hover:text-orange-400"
      href="#"
    >
      Instagram
    </a>
    <a
      class="mx-4 hover:text-orange-400"
      href="#"
    >
      Twitter
    </a>
  </div>
  <p
    class="mb-2"
  >
    © 2025 El Pizza. Все права защищены.
  </p>
  <p>
    Контакты:
     
    <a
      class="hover:text-orange-400"
      href="mailto:info@elpizza.com"
    >
      info@elpizza.com
    </a>
     
    | Телефон: +123 456 7890
  </p>
</footer>
`;

exports[`renders Footer component correctly 1`] = `
<footer
  class="bg-[rgb(228,228,228)] text-gray-800 py-6 text-center"
>
  <div
    class="social-icons flex justify-center mb-4"
  >
    <a
      class="mx-4 hover:text-orange-400"
      href="#"
    >
      Facebook
    </a>
    <a
      class="mx-4 hover:text-orange-400"
      href="#"
    >
      Instagram
    </a>
    <a
      class="mx-4 hover:text-orange-400"
      href="#"
    >
      Twitter
    </a>
  </div>
  <p
    class="mb-2"
  >
    © 2025 El Pizza. Все права защищены.
  </p>
  <p>
    Контакты:
     
    <a
      class="hover:text-orange-400"
      href="mailto:info@elpizza.com"
    >
      info@elpizza.com
    </a>
     
    | Телефон: +123 456 7890
  </p>
</footer>
`;
